src/Behat/Config/Config.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Behat\Config;
+
+final class Config
+{
+    public function __construct(
+        private array $settings = []
+    ) {
+    }
+
+    public function toArray(): array
+    {
+        return $this->settings;
+    }
```

Comment:
One small thought - do we want to / are we OK with enforcing that the PHP can only eventually return the same types as the YAML (e.g. only scalars, not callables etc)? I think that's probably useful to ensure `.php` / `.yaml` parity.

If so, then it *might* be worth making this class implement `JSONSerializable` instead of `->toArray()` and passing it through `json_decode(json_encode($config), true)` when we read it.

That way any future DTOs for subsections within the config would just have to implement `JSONSerializable` and we'd get recursive `->toArray()` "out of the box" rather than having to build that ourselves or users having to cast the sugar methods / objects back to array in their config files?

Reply:
Do we/you really want to keep yaml configuration forever?
Cause "callables" could be useful in the future even though I don't see any usage for now.

Reply:
My proposal is adding a ConfigInterface instead.

Like this one we have on Symfony
https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/Config/Builder/ConfigBuilderInterface.php

Reply:
I've added a commit with that proposal. Let me know what do you think about it.

Reply:
We do need to use yaml configuration at least until a new major version is launched. There will be many projects out there which will not want to update their config

Reply:
Yep, I agree, I a maintainer of Sylius packages which uses a lot Behat and I know that we need a major release for that. I was just suggesting the fact that we may not want to support the both configs forever. That's the reason I'm mitigated about limiting PHP configuration to only scalars.

Reply:
The `ConfigInterface` is good, though it doesn't quite solve my thought which was about supporting nested DTOs for partial config (e.g. the `GenericSuite` in your example on #1539, or if an extension wanted to provide one for their own config).

However, that isn't required for this, and on reflection I don't think tying to `JSONSerializable` is a good plan (that was probably the cough medicine talking :laughing:). So I'll carry on that bit of the conversation on #1539

Reply:
On the question of deprecating / dropping yaml support - I think actually if we're careful we could potentially do that quite soon without hurting people, again will pick up on #1539.